common/perf-test-core/src/main/java/com/azure/perf/test/core/ApiPerfTestBase.java

```diff
@@ -225,6 +234,103 @@ public Mono<Void> runAllAsync(long endNanoTime) {
             .then();
     }
 
+    public CompletableFuture<Void> runAllAsyncWithCompletableFuture(long endNanoTime) {
+        completedOperations = 0;
+        lastCompletionNanoTime = 0;
+        long startNanoTime = System.nanoTime();
+        Semaphore semaphore = new Semaphore(options.getConcurrentTaskLimit()); // Use configurable limit
+
+        List<CompletableFuture<Void>> futures = new ArrayList<>();
+
+        while (System.nanoTime() < endNanoTime) {
+            try {
+                semaphore.acquire();
+                // Each runTestAsyncWithCompletableFuture() call runs independently
+                CompletableFuture<Void> testFuture = runTestAsyncWithCompletableFuture()
+                    .thenAccept(result -> {
+                        completedOperations += result;
+                        lastCompletionNanoTime = System.nanoTime() - startNanoTime;
+                    })
+                    .whenComplete((res, ex) -> semaphore.release());
+                futures.add(testFuture);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException(e);
+            }
+        }
+
+        // Combine all futures so we can wait for all to complete
+        return CompletableFuture.allOf(futures.toArray(new CompletableFuture<?>[0]));
```

Comment:
May want to remove all `CompletableFuture`s from the list before passing it to `CompletableFuture.allOf`.

`futures.removeIf(CompletableFuture::isDone)`